### PR TITLE
TensorFlow DataSet Compatibility + other bug fixes

### PR DESCRIPTION
**Pull request by lucasrussell01**
- Introduced the possibility of reading data from previously saved tf datasets for training
- Added a script ROOT_to_tf to save ROOT files as a tf dataset
Minor fixes:
- Conversion from torch to tf in get_predict_generator fro apply training
- Change size-t -> long int in C++ parser to avoid problem observed when DeepTauVSjet cut set to -1
- Fixed small problem for file names with dashes for evaluation
- Made histo2d get weights shared pointer in definition
- Fix bug in regularisation implementation for convolutional layers
- Improved error messages when NaN found in tensor

### CODE REVIEW

File: Training/python/2018v1/Training_v0p1.py [hunk 1]
```diff
@@ -267,16 +290,38 @@ def compile_model(model, opt_name, learning_rate):
...
+            my_ds = ds
+        cell_locations = data_loader.cell_locations
+        active_features = data_loader.active_features
+        active = [] #list of elements to be kept
+        if "TauFlat" in active_features:
+            active.append(1)
+        if "inner" in cell_locations:
+            active.extend([2,3,4])
+        if "outer" in cell_locations:
+            active.extend([5,6,7])
```
Comment (shedprog) [inline Training/python/2018v1/Training_v0p1.py:308]: The order of the cell may not be static and depends on how we define it in config: https://github.com/cms-tau-pog/TauMLTools/blob/ed7b7b385219fa6a239bf972502cb8edf92754d1/Training/configs/training_v1.yaml#L72-L76
Better to read dynamically order of CellType from config or precompiled enume classes. Same corresponds to rm_inner function.

File: Training/python/DataLoader.py [hunk 2]
```diff
@@ -118,6 +121,8 @@ def get_generator(self, primary_set = True, return_truth = True, return_weights
         def _generator():
 
             finish_counter = 0
+            if show_progress:
+                pbar = tqdm(total = n_batches)
```
Comment (shedprog) [inline Training/python/DataLoader.py:125]: `n_batches` is not always a reasonable number, e.g: It is set to -1 when we are reading all dataset.

File: Training/python/DataLoader.py [hunk 3]
```diff
@@ -164,6 +171,7 @@ def get_predict_generator(self):
         '''
         assert self.batch_size == 1
         data_loader = R.DataLoader()
+        converter = torch_to_tf(return_truth=True, return_weights=False)
```
Comment (shedprog) [inline Training/python/DataLoader.py:174]: I propose to add this args to get_predict_generator signature and read it from there

File: Training/interface/DataLoader_main.h [hunk 4]
```diff
@@ -190,10 +190,10 @@ class DataLoader {
         input_histogram .th2d_add(*(input_th2d .get()));
 
         target_histogram.divide(input_histogram);
-        hist_weights[tau_type] = std::make_shared<TH2D>(target_histogram.get_weights_th2d(
+        hist_weights[tau_type] = target_histogram.get_weights_th2d(
```
Comment (shedprog) [inline Training/interface/DataLoader_main.h:193]: Is this change needed because of the scope?

**Review by shedprog — CHANGES_REQUESTED** [review]

File: Training/python/2018v1/ROOT_to_tf.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+dataloader = DataLoader.DataLoader(training_cfg, scaling_cfg)
+print("DataLoader Created")
+gen_train = dataloader.get_generator(primary_set = True, return_weights = dataloader.use_weights, show_progress=True) 
+print("Generator Loaded")
+input_shape, input_types = dataloader.get_input_config()
+print("Input shapes and Types acquired")
+data_train = tf.data.Dataset.from_generator(gen_train, output_types = input_types, output_shapes = input_shape).prefetch(tf.data.AUTOTUNE)
+print("Dataset extracted from DataLoader")
+tf.data.experimental.save(data_train, save_path, compression = "GZIP")
```
[the remark this answers is not in the document]
Comment (lucasrussell01) [reply Training/python/2018v1/ROOT_to_tf.py:46]: There's no option to limit the memory of this function, but it saves the files continuously, it doesn't wait until everything is loaded.

File: Training/python/DataLoader.py (hunk 2 above)
Comment (lucasrussell01) [reply Training/python/DataLoader.py:125]: Will implement this thanks, it's just a progress bar for when loading the dataset using tensorflow, so will disable it if n_batches<0

File: Training/python/DataLoader.py (hunk 3 above)
Comment (lucasrussell01) [reply Training/python/DataLoader.py:174]: Will also add this

File: Training/interface/DataLoader_main.h (hunk 4 above)
Comment (lucasrussell01) [reply Training/interface/DataLoader_main.h:193]: Yes

File: Training/python/2018v1/Training_v0p1.py (hunk 1 above)
Comment (lucasrussell01) [reply Training/python/2018v1/Training_v0p1.py]: @shedprog I've now added an argument tf_dataset_x_order, which will allow us to specify the order of the layers in the tf dataset we load, in case eg we disabled some when it was generated, and rm_inner and reshape_tensor now read the relevant indeces from this

File: Training/python/2018v1/Training_v0p1.py [hunk 6]
```diff
@@ -267,16 +295,42 @@ def compile_model(model, opt_name, learning_rate):
...
+            my_ds = ds
+        cell_locations = data_loader.cell_locations
+        active_features = data_loader.active_features
+        active = [] #list of elements to be kept
+        if "TauFlat" in active_features:
+            active.append(tauflat_index)
+        if "inner" in cell_locations:
+            active.extend(inner_indices)
+        if "outer" in cell_locations:
+            active.extend(outer_indices)
```
Comment (shedprog) [inline Training/python/2018v1/Training_v0p1.py:317]: To my understanding the model input shape is pre-defined in the `netConf_full` variable, you can not just change the number of tensors in the input (to NN) tuple because it already considers fixed number and defined here: https://github.com/cms-tau-pog/TauMLTools/blob/ed7b7b385219fa6a239bf972502cb8edf92754d1/Training/python/DataLoader.py#L185-L219

Also my question related to the indexes of the tensor (em, muon, hadrons,...) was referring that the sequential number of this index is already defined in meta enume class (see attachments) [SetupMeta.txt](https://github.com/cms-tau-pog/TauMLTools/files/8237831/SetupMeta.txt)
if you at some point need this index, maybe it would be better by designed not to define `tf_dataset_x_order` but take it from `enum class CellObjectType {` for instance `getattr(R.CellObjectType,"PfCand_electron")` would give you 1 or if you need the order of the grouped tensors after DataLoading it is following this order: https://github.com/cms-tau-pog/TauMLTools/blob/ed7b7b385219fa6a239bf972502cb8edf92754d1/Training/configs/training_v1.yaml#L72-L76 if you remove from there one of the list it should not return this tensor. But in principle this part is not interfering with RUN3 training, so it is up to you.

Comment (lucasrussell01) [reply Training/python/2018v1/Training_v0p1.py:317]: From the testing I've done, changing the number of inputs in the tensors works if you change the parameters that I introduced to switch off either inner or outer cones (cell_locations) or the tau dense layer (removing it from active_features), this information is passed to net config. We then need to remove the concerned layers from the saved tensor if using tf input format as they are usually created with all layers active (eg generated with all 7 tensors 3xinner, 3xouter and tau dense, so if we want to then disable for example inner for a training from this tf dataset we need to remove the 3 inner tensors from the dataset as the model will only expect 4 tensors). Let me know if this doesn't make sense.

Comment (lucasrussell01) [reply Training/python/2018v1/Training_v0p1.py:317]: Will also have a look at the indexes you're referring to in SetupMeta, thanks!

File: Training/interface/DataLoader_main.h (hunk 4 above)
Comment (kandrosov) [reply Training/interface/DataLoader_main.h:193]: the problem is that [get_weigths_th2d](https://github.com/cms-tau-pog/TauMLTools/blob/ed7b7b385219fa6a239bf972502cb8edf92754d1/Training/interface/histogram2d.h#L176) (as it is in master) created shared pointer inside, but returned a reference:
```c++
TH2D& Histogram_2D::get_weights_th2d(const char* name, const char* title) {
  ...
  histo_ = std::shared_ptr<TH2D>(new TH2D(name, title, nx, xmin_, xmax_, ny, yaxis_.front(), yaxis_.back()));
  return *histo_;
}
```

So the original `histo_` was destructed before the copy constructor is called...

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Comment by lucasrussell01** [comment]
@shedprog No I have not yet found a fix for that, this was just a typo I happened to find

File: Training/python/2018v1/Training_v0p1.py [hunk 7]
```diff
@@ -303,7 +303,19 @@ def run_training(model, data_loader, to_profile, log_suffix):
...
+            if n_inner % 2 == 0 or n_outer % 2 == 0:
+                raise Exception("Number of cells not supported")
+            inner_size = data_loader.inner_cell_size
+            outer_size = data_loader.outer_cell_size
+            n_inner_right = (n_inner - 1) / 2
+            n_outer_right = np.ceil(n_inner_right * inner_size /  outer_size)
+            n_outer_to_remove = n_outer_right*2 + 1
+            i_middle = (n_outer-1)/2
+            i_start = int(i_middle - (n_outer_to_remove-1)/2)
+            i_end = int(i_middle + (n_outer_to_remove-1)/2 + 1) # +1 as end index not included
```
Comment (kandrosov) [inline Training/python/2018v1/Training_v0p1.py:317]: `(n_outer_to_remove-1)/2 == n_outer_right` => remove `n_outer_to_remove` and use `n_outer_right`

Comment (lucasrussell01) [reply Training/python/2018v1/Training_v0p1.py]: @kandrosov Good point, fixed now
